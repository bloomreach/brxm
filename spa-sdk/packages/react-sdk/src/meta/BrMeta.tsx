/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaCollection } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: MetaCollection;
}

export class BrMeta extends React.Component<BrMetaProps> {
  private headRef = React.createRef<HTMLElement>();

  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.renderMeta();
  }

  componentDidUpdate(prevProps: BrMetaProps) {
    prevProps.meta.clear();

    this.renderMeta();
  }

  componentWillUnmount() {
    this.props.meta.clear();
  }

  private renderMeta() {
    const head = this.headRef?.current?.nextSibling;
    const tail = this.tailRef?.current;

    if (!head || !tail) {
      return;
    }

    this.props.meta.render(head, tail);
  }

  render() {
    return (
      <>
        {this.props.meta.length > 0 && <span style={{ display: 'none' }} ref={this.headRef} />}
        {this.props.children}
        {this.props.meta.length > 0 && <span style={{ display: 'none' }} ref={this.tailRef} />}
      </>
    );
  }
}
