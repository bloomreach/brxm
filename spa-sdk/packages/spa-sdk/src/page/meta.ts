/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const TYPE_META_COMMENT = 'comment';

/**
 * Meta-data following before a page component.
 */
export const META_POSITION_BEGIN = 'begin';

/**
 * Meta-data following after a page component.
 */
export const META_POSITION_END = 'end';

export type MetaType = typeof TYPE_META_COMMENT;
export type MetaPosition = typeof META_POSITION_BEGIN | typeof META_POSITION_END;

export interface MetaModel {
  data: string;
  type: MetaType;
}

/**
 * Meta information describing a part of the page.
 */
export interface Meta {
  /**
   * @return The meta-data.
   */
  getData(): string;

  /**
   * @return The meta-data position relative to the related element.
   */
  getPosition(): MetaPosition;
}

export class MetaImpl implements Meta {
  constructor(protected model: MetaModel, protected position: MetaPosition) {}

  getData() {
    return this.model.data;
  }

  getPosition() {
    return this.position;
  }
}

/**
 * Checks whether a value is a meta-data object.
 * @param value The value to check.
 */
export function isMeta(value: any): value is Meta {
  return value instanceof MetaImpl;
}
