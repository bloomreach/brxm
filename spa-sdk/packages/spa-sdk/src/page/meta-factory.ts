/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { injectable } from 'inversify';
import { SimpleFactory } from './factory';
import { MetaModel, MetaPosition, MetaType, Meta } from './meta';

type MetaBuilder = (model: MetaModel, position: MetaPosition) => Meta;

/**
 * The factory to produce meta-data collection from the page model meta-data.
 */
@injectable()
export class MetaFactory extends SimpleFactory<MetaType, MetaBuilder> {
  create(meta: MetaModel, position: MetaPosition) {
    const builder = this.mapping.get(meta.type);
    if (!builder) {
      throw new Error(`Unsupported meta type: '${meta.type}'.`);
    }

    return builder(meta, position);
  }
}
