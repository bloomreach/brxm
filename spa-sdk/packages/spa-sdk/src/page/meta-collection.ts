/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { MetaFactory } from './meta-factory';
import { MetaModel, Meta, META_POSITION_BEGIN, META_POSITION_END } from './meta';
import { isMetaComment } from './meta-comment';

export const MetaCollectionModelToken = Symbol.for('MetaCollectionModelToken');

export interface MetaCollectionModel {
  beginNodeSpan?: MetaModel[];
  endNodeSpan?: MetaModel[];
}

/**
 * Collection of the meta-data describing a part of the page.
 * @note The collection extends the built-in Array type for backward compatibility.
 */
export interface MetaCollection extends Array<Meta> {
  /**
   * Clears all previously rendered meta-data objects.
   * @deprecated Use a callback returned by the `render` method.
   */
  clear(): void;

  /**
   * Renders meta-data objects on the page.
   * @param head The heading node of the page fragment.
   * @param tail The tailing node of the page fragment.
   * @return The callback to clear rendered meta-data objects.
   */
  render(head: Node, tail: Node): () => void;
}

@injectable()
export class MetaCollectionImpl extends Array<Meta> implements MetaCollection {
  private comments: Comment[] = [];

  constructor(
    @inject(MetaCollectionModelToken) model: MetaCollectionModel,
    @inject(MetaFactory) metaFactory: MetaFactory,
  ) {
    super(
      ...(model.beginNodeSpan || []).map(model => metaFactory.create(model, META_POSITION_BEGIN)),
      ...(model.endNodeSpan || []).map(model => metaFactory.create(model, META_POSITION_END)),
    );

    const prototype = Object.create(MetaCollectionImpl.prototype);

    prototype.constructor = Array.prototype.constructor;
    Object.setPrototypeOf(this, prototype);
    Object.freeze(this);
  }

  clear(comments = [...this.comments]) {
    comments.forEach((comment) => {
      comment.remove();

      const index = this.comments.indexOf(comment);
      if (index > -1) {
        this.comments.splice(index, 1);
      }
    });
  }

  render(head: Node, tail: Node) {
    const document = head.ownerDocument ?? tail.ownerDocument;
    const comments = document
      ? [
        ...this.filter(isMetaComment)
          .filter(meta => meta.getPosition() === META_POSITION_BEGIN)
          .map(meta => document.createComment(meta.getData()))
          .map((comment) => {
            head.parentNode?.insertBefore(comment, head);

            return comment;
          }),

        ...this.filter(isMetaComment)
          .filter(meta => meta.getPosition() === META_POSITION_END)
          .reverse()
          .map(meta => document.createComment(meta.getData()))
          .map((comment) => {
            if (tail.nextSibling) {
              tail.parentNode?.insertBefore(comment, tail.nextSibling);
            } else {
              tail.parentNode?.appendChild(comment);
            }

            return comment;
          }),
      ]
      : [];

    this.comments.push(...comments);

    return this.clear.bind(this, comments);
  }
}
