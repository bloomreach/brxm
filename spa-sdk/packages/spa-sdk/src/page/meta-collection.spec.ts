/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isMetaCollection, MetaCollectionImpl, MetaCollectionModel } from './meta-collection';
import { MetaFactory } from './meta-factory';
import { MetaType, Meta, META_POSITION_BEGIN, META_POSITION_END, TYPE_META_COMMENT } from './meta';
import { MetaCommentImpl } from './meta-comment';

let factory: jest.Mocked<MetaFactory>;

beforeEach(() => {
  factory = { create: jest.fn() } as unknown as jest.Mocked<MetaFactory>;
});

function createMetaCollection(model = {} as MetaCollectionModel) {
  return new MetaCollectionImpl(model, factory);
}

describe('MetaCollectionImpl', () => {
  describe('constructor', () => {
    it('should extend a built-in array class', () => {
      const collection = createMetaCollection();

      expect(collection).toBeInstanceOf(Array);
    });

    it('should call a factory to build meta-data', () => {
      createMetaCollection({
        beginNodeSpan: [
          { data: 'meta1', type: 'type1' as MetaType },
          { data: 'meta2', type: 'type2' as MetaType },
        ],
      });

      expect(factory.create).toBeCalledWith({ data: 'meta1', type: 'type1' }, expect.anything());
      expect(factory.create).toBeCalledWith({ data: 'meta2', type: 'type2' }, expect.anything());
    });

    it('should pass a position of the meta', () => {
      createMetaCollection({
        beginNodeSpan: [{ data: 'meta1', type: 'type1' as MetaType }],
        endNodeSpan: [{ data: 'meta2', type: 'type2' as MetaType }],
      });

      expect(factory.create).toBeCalledWith({ data: 'meta1', type: 'type1' }, META_POSITION_BEGIN);
      expect(factory.create).toBeCalledWith({ data: 'meta2', type: 'type2' }, META_POSITION_END);
    });

    it('should hold meta-data items', () => {
      const meta1 = { data: 'meta1' } as unknown as Meta;
      const meta2 = { data: 'meta2' } as unknown as Meta;

      factory.create.mockReturnValueOnce(meta1);
      factory.create.mockReturnValueOnce(meta2);

      const collection = createMetaCollection({
        beginNodeSpan: [{ data: 'meta1', type: 'type1' as MetaType }],
        endNodeSpan: [{ data: 'meta2', type: 'type2' as MetaType }],
      });

      expect(collection).toMatchSnapshot();
    });
  });

  describe('clear', () => {
    let a: HTMLElement;
    let collection: MetaCollectionImpl;

    beforeEach(() => {
      collection = new MetaCollectionImpl(
        {
          beginNodeSpan: [{ data: 'meta1', type: TYPE_META_COMMENT }],
          endNodeSpan: [{ data: 'meta2', type: TYPE_META_COMMENT }],
        },
        new MetaFactory()
          .register(TYPE_META_COMMENT, (model, position) => new MetaCommentImpl(model, position))
      );

      a = document.createElement('a');

      document.body.append(a);
    });

    afterEach(() => {
      document.body.innerHTML = '';
    });

    it('should remove rendered comments', () => {
      collection.render(a, a);
      collection.clear();

      expect(document.body).toMatchSnapshot();
    });
  });

  describe('render', () => {
    let a: HTMLElement;
    let b: HTMLElement;
    let collection: MetaCollectionImpl;

    beforeEach(() => {
      collection = new MetaCollectionImpl(
        {
          beginNodeSpan: [
            { data: 'meta1', type: TYPE_META_COMMENT },
            { data: 'meta2', type: TYPE_META_COMMENT },
          ],
          endNodeSpan: [
            { data: 'meta3', type: TYPE_META_COMMENT },
            { data: 'meta4', type: TYPE_META_COMMENT },
          ],
        },
        new MetaFactory()
          .register(TYPE_META_COMMENT, (model, position) => new MetaCommentImpl(model, position)),
      );

      a = document.createElement('a');
      b = document.createElement('b');

      document.body.append(a, b);
    });

    afterEach(() => {
      document.body.innerHTML = '';
    });

    it('should not render on detached element', () => {
      document.body.removeChild(a);

      collection.render(a, a);

      expect(document.body).toMatchSnapshot();
    });

    it('should not render on document node', () => {
      const node = document.body.parentNode?.parentNode as Node;

      collection.render(node, node);

      expect(document.body).toMatchSnapshot();
    });

    it('should surround an element with comments', () => {
      collection.render(a, a);

      expect(document.body).toMatchSnapshot();
    });

    it('should surround multiple elements with comments', () => {
      collection.render(a, b);

      expect(document.body).toMatchSnapshot();
    });

    it('should return a callback clearing rendered comments', () => {
      const clear = collection.render(a, a);
      clear();

      expect(clear).toBeInstanceOf(Function);
      expect(document.body).toMatchSnapshot();
    });

    it('should return a callback clearing only rendered comments during the call', () => {
      const clear = collection.render(a, a);
      collection.render(b, b);
      clear();

      expect(document.body).toMatchSnapshot();
    });
  });
});

describe('isMetaCollection', () => {
  it('should return true', () => {
    const meta = createMetaCollection();

    expect(isMetaCollection(meta)).toBe(true);
  });

  it('should return false', () => {
    expect(isMetaCollection(undefined)).toBe(false);
    expect(isMetaCollection([])).toBe(false);
    expect(isMetaCollection({})).toBe(false);
  });
});
