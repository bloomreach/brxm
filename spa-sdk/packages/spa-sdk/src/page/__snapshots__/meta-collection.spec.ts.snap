// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaCollectionImpl clear should remove rendered comments 1`] = `
<body>
  <a />
</body>
`;

exports[`MetaCollectionImpl constructor should hold meta-data items 1`] = `
Array [
  Object {
    "data": "meta1",
  },
  Object {
    "data": "meta2",
  },
]
`;

exports[`MetaCollectionImpl render should not render on detached element 1`] = `
<body>
  <b />
</body>
`;

exports[`MetaCollectionImpl render should not render on document node 1`] = `
<body>
  <a />
  <b />
</body>
`;

exports[`MetaCollectionImpl render should surround an element with comments 1`] = `
<body>
  <!--meta1-->
  <!--meta2-->
  <a />
  <!--meta3-->
  <!--meta4-->
  <b />
</body>
`;

exports[`MetaCollectionImpl render should surround multiple elements with comments 1`] = `
<body>
  <!--meta1-->
  <!--meta2-->
  <a />
  <b />
  <!--meta3-->
  <!--meta4-->
</body>
`;
